### DISCUSSION TITLE
When paired with a custom server setPreviewData is not working

### DISCUSSION THREAD

**Discussion by cidicles** [vercel/next.js#16766] [Help] [answered]
# Bug report

## Describe the bug

When using a [custom server](https://nextjs.org/docs/advanced-features/custom-server) the `setPreviewData` function is not setting the cookies as expected. If `next start` or `next dev` is used the cookies will be set. 

## To Reproduce

1. Create a next site using a [custom server](https://nextjs.org/docs/advanced-features/custom-server).
2. Follow the instructions for [preview mode](https://nextjs.org/docs/advanced-features/preview-mode)
3. Visit the `/api/preview` route

## Expected behavior

The `__next_preview_data` and `__prerender_bypass` cookies are set. 

## System information

- OS: macOS
- Version of Next.js: 9.5.3
- Version of Node.js: 10.15.0

**Comment by SamStenton** [comment, marked as answer]
Hey so I had this same issue when using express as my custom server on localhost. When I inspected the request that sets `setPreviewData` it was complaining that the cookie not being set over HTTPS. 

<img width="742" alt="Screenshot 2020-09-08 at 15 54 58" src="https://user-images.githubusercontent.com/5670870/92492828-b7b62e80-f1eb-11ea-8cb1-2e3ffe7a1a55.png">

**Workaround**: 

```javascript

...

res.setPreviewData({})

// Remove secure requirement for Cookies if on localhost
if(process.env.NODE_ENV === 'development') {
	const cookies = res.getHeader('Set-Cookie');
  	res.setHeader('Set-Cookie', cookies.map((cookie) => cookie.replace('Secure;', '')))
}

...
```

I'm going to dig into a cleaner way of handling this. 

**Edit**

Looks like this is an issue with https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/api-utils.ts#L473

Logging out the value of `process.env.NODE_ENV` outputs `undefined`

**Reply by ambrauer** [reply]
Just ran into this one as well. In my case for the workaround, I also had to set the `SameSite` cookie to `Lax`.

```js
  if(process.env.NODE_ENV === 'development') {
    const cookies = res.getHeader('Set-Cookie') as string[];
    res.setHeader('Set-Cookie', cookies.map((cookie) => cookie.replace('Secure;', '').replace('SameSite=None', 'SameSite=Lax')))
  }
```
